grift/src/GRIFT/Simulation.hs

```diff
@@ -152,16 +152,25 @@ evalInstExpr :: forall m rv fmt w . (RVStateM m rv, KnownRV rv)
              -> Instruction rv fmt -- ^ instruction
              -> Integer            -- ^ Instruction width (in bytes)
              -> InstExpr fmt rv w  -- ^ Expression to be evaluated
-             -> m (BitVector w)
-evalInstExpr _ _ _ (InstLitBV bv) = return bv
-evalInstExpr iset inst ib (InstAbbrevApp abbrevApp) = evalInstExpr iset inst ib (expandAbbrevApp abbrevApp)
-evalInstExpr _ (Inst _ (Operands _ operands)) _ (OperandExpr _ (OperandID p)) = return (operands !! p)
+             -> m (BV w)
+evalInstExpr _ _ _ (InstLitBV _ bv) = return bv
+evalInstExpr iset inst ib (InstAbbrevApp abbrevApp) =
+  evalInstExpr iset inst ib (expandAbbrevApp abbrevApp)
+evalInstExpr _ (Inst _ (Operands _ operands)) _ (OperandExpr _ (OperandID p)) =
+  return (getSizedBV (operands !! p))
 evalInstExpr _ _ ib (InstBytes _) = do
   rv <- getRV
-  return $ withRV rv $ bitVector ib
-evalInstExpr iset inst _ (InstWord _) = do
-  rv <- getRV
-  return $ (withRV rv $ bvZext $ encode iset inst)
+  return $ withRV rv $ mkBV' ib
+evalInstExpr iset inst _ (InstWord _) =
+  -- TODO: this is *very* awkward, because we z-extend, which requires '32 <=
+  -- w', but adding the constraint makes it akward when 'evalInstExpr' is passed
+  -- to 'evalStateApp' in the next case.
+  case testLeq (knownNat @32) (knownNat @w) of
+    Just LeqProof ->
+      do
+        rv <- getRV
+        return $ withRV rv $ zextOrId $ getSizedBV $ encode iset inst
```

Comment:
This workaround feels wrong.

The problem is, because of the z-extension here, we want to know that `32 <= w` holds.

But if we add this constraint, because of the call to `evalStateApp` on line 174, we must also change the type of `evalStateApp` to have a `32 <= w` in its callback constraint, which gets even more weird, because now we'd need `evalLocApp`, etc. to also have such a constraint...

We should try and find where exactly expressions need to be `32 <= w` and where there should be a cut-off.